### PR TITLE
Update onINP to also observe first-input entries

### LINKED ISSUE #228: Consider expose FID `duration` as first `INP` score, for cases when its under `durationThreshold`
`onINP` relies on event timing `durationThreshold` and provides a default of 40ms (configurable down to 16ms).

There are cases where we report `onFID` but not `onINP` because of this threshold.  `first-input` event timing actually does already expose the `duration` value so we can report the full latency exactly as INP would measure it, and not just the input delay.

Would help address https://github.com/GoogleChrome/web-vitals-extension/issues/103

### PR DESCRIPTION
**Pull request by philipwalton**
Fixes #228 by having the observer created in `onINP()` also observe `first-input` entries (in addition to `event`)—given that they are both instances of `PerformanceEventTiming`.

This PR also includes logic to remove duplicate interactions by checking to see if a `first-input` and `event` entry both have the same `startTime` and `duration` values. This logic is needed until [crbug/1325826](https://crbug.com/1325826) is fixed in Chrome, which adds support for the `interactionId` property on `first-input` entries.

/cc @mmocny

### CODE REVIEW

File: src/onINP.ts [hunk 1]
```diff
@@ -119,6 +119,26 @@ export const onINP = (onReport: ReportCallback, opts?: ReportOpts) => {
...
         processEntry(entry);
       }
+
+      // Entries of type `first-input` don't currently have an `interactionId`,
+      // so to consider them in INP we have to first check that an existing
+      // entry doesn't match the `duration` and `startTime`.
+      // Note that this logic assumes that `event` entries are dispatched
+      // before `first-input` entries. This is true in Chrome but it is not
+      // true in Firefox; however, Firefox doesn't support interactionId, so
+      // it's not an issue at the moment.
```
Comment (mmocny) [inline src/onINP.ts:129]: I was surprised to learn this, since I think we fire FID first.

However, I guess you mean that the order of Entries in the PO callbacks, when there are multiple observe() calls of different types... This I haven't tested and I'm not sure if order is guaranteed or arbitrary right now.

File: src/onINP.ts [hunk 2]
```diff
@@ -119,6 +119,26 @@ export const onINP = (onReport: ReportCallback, opts?: ReportOpts) => {
...
+      // TODO(philipwalton): remove once crbug.com/1325826 is fixed.
+      if (entry.entryType === 'first-input') {
+        const noMatchingEntry = !longestInteractionList.some((interaction) => {
+          return interaction.entries.some((prevEntry) => {
+            return entry.duration === prevEntry.duration &&
+                entry.startTime === prevEntry.startTime;
+          });
+        });
+        if (noMatchingEntry) {
+          processEntry(entry);
```
Comment (mmocny) [inline src/onINP.ts:139]: What happens if:
- FID is pointerdown, but we do not get a pointerdown event timing entry
- Then we do get an pointerup event timing entry (for the same interaction)

I think it will be treated as 2 different interactions here, also affecting p98.

File: src/onINP.ts [hunk 3]
```diff
@@ -143,6 +163,10 @@ export const onINP = (onReport: ReportCallback, opts?: ReportOpts) => {
   report = bindReporter(onReport, metric, opts.reportAllChanges);
 
   if (po) {
+    // Also observe entries of type `first-input`. This is useful in cases
+    // where the first interaction is less than the `durationThreshold`.
+    po.observe({type: 'first-input', buffered: true});
```
Comment (mmocny) [inline src/onINP.ts:168]: Yeah this seems better :)

File: test/views/inp.njk [hunk 4]
```diff
@@ -89,7 +89,7 @@
       inp.entries = inp.entries.map((e) => ({
         ...e.toJSON(),
         interactionId: e.interactionId,
-        target: e.target.nodeName.toLowerCase(),
+        target: e.target?.nodeName.toLowerCase(),
```
Comment (mmocny) [inline test/views/inp.njk:92]: Is this related to FID entry?

**Review by mmocny — APPROVED** [review]
I don't think this is perfect either, but I think its a good change and cleaner than the one I suggested.

Cheers!

File: src/onINP.ts (hunk 1 above)
Comment (philipwalton) [reply src/onINP.ts:129]: Yeah, I only tested this in Chrome on Mac (desktop), but no matter what I did, which order I called `observe()` in, or which event I used, I always got the `event` entry processed before the `first-input` entry.

File: test/views/inp.njk (hunk 4 above)
Comment (philipwalton) [reply test/views/inp.njk:92]: Yeah, not sure why but I started noticing lots of cases where `target` was undefined on `first-input` entries. I didn't look much further into repo-ing why.

File: src/onINP.ts (hunk 2 above)
Comment (philipwalton) [reply src/onINP.ts:139]: Yeah, I think so. My hope is we can remove this logic entirely once `first-input` entries have an `interactionId` property. My goal here was to minimize these types of occurrences .
